### PR TITLE
Replace tab size link mini dlg with edit control

### PR DESCRIPTION
**Pull request by ozone10**
image

fix https://github.com/notepad-plus-plus/notepad-plus-plus/issues/11695
fix https://github.com/notepad-plus-plus/notepad-plus-plus/issues/13176

### CODE REVIEW

**Comment by Yaron10** [comment]
@ozone10,

👍 
Thank you for this too. Two more. :)

1. If the user enters `0` (zero) in the EditBox and moves focus to another element, that value is not auto-changed.
2. `Replace with Spaces` might still be blurry in DarkMode .

**Comment by Yaron10** [comment]
@ozone10,

Thank you for the quick fix. 👍 

How about implementing it in `if (tabSize < 1)` under `case EN_CHANGE:`?
That is: emulate Windows behavior on trying to enter non-numeric chars, and if the users enters `0` - beep and restore the previous value.
(The user won't be able to type `01`, but that shouldn't be a problem IMO).

**Comment by ozone10** [comment]
@Yaron10 

> 2. `Replace with Spaces` might still be blurry in DarkMode .

That is checkbox button text, and use different text drawing for disabled state than static text, if there was issue with it, it should also affect other places.

> How about implementing it in `if (tabSize < 1)` under `case EN_CHANGE:`? That is: emulate Windows behavior on trying to enter non-numeric chars, and if the users enters `0` - beep and restore the previous value. (The user won't be able to type `01`, but that shouldn't be a problem IMO).

Currently there is no difference between `0` and empty value. Differentiating would make code more complex.
Also consider value e.g. `10` user will put caret before `1` press `Delete` -> value will change to `0` -> value will be restored to `10`.  This is not good.

**Comment by Yaron10** [comment]
@ozone10,

> Currently there is no difference between 0 and empty value.
...
Also consider value e.g. 10 user will put caret before 1...

Both arguments are very good. Thank you for the explanation. 👍 

***

תמונה
תמונה

If you can't reproduce it, it may be caused by my Windows 10 settings and custom theme.
You know I don't use DarkMode. :)

File: PowerEditor/src/WinControls/Window.h [hunk 1]
```diff
@@ -64,8 +64,20 @@ class Window
 			::UpdateWindow(_hSelf);
 	}
 
+	virtual void redrawDlgItem(const int nIDDlgItem, bool forceUpdate = false) const
```
Comment (donho) [inline PowerEditor/src/WinControls/Window.h:67]: What does it do exactly this function?
And why do we need it?

**Review by donho — CHANGES_REQUESTED** [review]
`redrawDlgItem` uses MapWindowPoints function which converts (maps) a set of points from a coordinate space relative to one window to a coordinate space relative to another window, then redraw it with the new coordonate points.
I don't see the reason for it.
Please provide the arguments for this new added method.

File: PowerEditor/src/WinControls/Window.h (hunk 1 above)
Comment (ozone10) [reply PowerEditor/src/WinControls/Window.h:67]: @donho 
currently using `redraw()` will redraw all items in window which is unnecessary and can cause flickering.
Try switching fast options in Preference->Editing->Current Line Indicator.

`MapWindowPoints` is used to get correct coordinates of child rect for redrawing.
https://stackoverflow.com/questions/18034975/how-do-i-find-position-of-a-win32-control-window-relative-to-its-parent-window

Comment (donho) [reply PowerEditor/src/WinControls/Window.h:67]: @ozone10 
In this case, this method should be in `class StaticDialog` instead of `class Window`.

**Review by donho — CHANGES_REQUESTED** [review]
Please put the method redrawDlgItem in `class StaticDialog` instead of `class Window`.

**Comment by Yaron10** [comment]
Fixes also https://github.com/notepad-plus-plus/notepad-plus-plus/issues/12558.
👍

**Pull request closed without merging**